py/dml/c_backend.py

```diff
@@ -2258,10 +2307,118 @@ def generate_alloc_trait_vtables(node):
         if subnode.traits:
             for trait in subnode.traits.referenced:
                 name = subnode.traits.vtable_cname(trait)
-                out(f'{name} = MM_MALLOC(1, typeof(*{name}));\n')
+                out(f'{name} = MM_ZALLOC(1, typeof(*{name}));\n')
     out('}\n', preindent=-1)
     splitting_point()
 
+def generate_startup_calls(devnode):
+    start_function_definition('void _startup_calls(void)')
+    out('{\n', postindent=1)
+
+    def startup_calls(startup_methods):
+        by_dims = {}
+        for (method_data_tuple, dims) in startup_methods:
+            by_dims.setdefault(dims, []).append(method_data_tuple)
+        all_dims = sorted(by_dims)
+        prev_dims = ()
+        for dims in all_dims:
+            for (site, gen_call, data) in by_dims[dims]:
+                common_dims = []
+                for (a, b) in zip(dims, prev_dims):
+                    if a == b:
+                        common_dims.append(a)
+                    else:
+                        break
+                for _ in range(len(common_dims), len(prev_dims)):
+                    out('}\n', preindent=-1)
+                for i in range(len(common_dims), len(dims)):
+                    idxvar = '_i%d' % (i,)
+                    out('for (uint32 %s = 0; %s < %d; %s++) {\n'
+                        % (idxvar, idxvar, dims[i], idxvar),
+                        postindent=1)
+                prev_dims = dims
+
+                idxvars = ['_i%d' % (i,) for i in range(len(dims))]
+                gen_call(data, idxvars)
+
+        for _ in prev_dims:
+            out('}\n', preindent=-1)
+
+    def generate_regular_call(method, idxvars):
+        site = method.site
+        indices = tuple(mkLit(site, idx, TInt(32, False)) for idx in idxvars)
+        outargs = [mkLit(site,
+                         ('*((%s) {0})'
+                          % ((TArray(t, mkIntegerLiteral(site, 1))
+                              .declaration('')),)),
+                         t)
+                   for (_, t) in method.outp]
+        with IgnoreFailure(site):
```

Comment:
why IgnoreFailure? Shouldn't startup be nonthrowing, so that NoFailure is sufficient? (same goes for generate_trait_calls)

Reply:
It makes sense when you view throwing as an output parameter. Throws are ignored during startup, but an independent startup memoized can be called in order to query whether or not it threw during startup.
Non-memoized independent startups are not allowed to throw.

Reply:
Ah, I see. Perhaps a comment to explain that?

(I find memoized throws a bit odd to start with, but I can't deny that they add some symmetry, so I choose not to complain about them)